examples/apps/tree-comparison/src/model/legacyTreeInventoryList.ts

```diff
@@ -0,0 +1,244 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	BuildNode,
+	Change,
+	EagerCheckout,
+	SharedTree as LegacySharedTree,
+	NodeId,
+	StablePlace,
+	StableRange,
+	TraitLabel,
+	TreeView,
+	TreeViewNode,
+} from "@fluid-experimental/tree";
+import { DataObject, DataObjectFactory } from "@fluidframework/aqueduct";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+
+import type { IInventoryItem, IInventoryList } from "../modelInterfaces";
+import { InventoryItem } from "./inventoryItem";
+
+const legacySharedTreeKey = "legacySharedTree";
+
+// TODO: Do I want string here or number?  Prob don't want anything tree-specific
+// since this gets exposed to the view.  I've currently removed any reverse lookups so it doesn't
+// matter too much.
+const nodeIdToInventoryItemId = (nodeId: NodeId) => nodeId.toString();
+
+export class LegacyTreeInventoryList extends DataObject implements IInventoryList {
+	private _tree: LegacySharedTree | undefined;
+	private readonly _inventoryItems = new Map<string, InventoryItem>();
+
+	private get tree() {
+		if (this._tree === undefined) {
+			throw new Error("Not initialized properly");
+		}
+		return this._tree;
+	}
+
+	public readonly addItem = (name: string, quantity: number) => {
+		const addedNode: BuildNode = {
+			definition: "inventoryItem",
+			traits: {
+				name: {
+					definition: "name",
+					payload: name,
+				},
+				quantity: {
+					definition: "quantity",
+					payload: quantity,
+				},
+			},
+		};
+		const rootNode = this.tree.currentView.getViewNode(this.tree.currentView.root);
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const inventoryNodeId = rootNode.traits.get("inventory" as TraitLabel)![0];
+		this.tree.applyEdit(
+			Change.insertTree(
+				addedNode,
+				StablePlace.atEndOf({
+					parent: inventoryNodeId,
+					label: "inventoryItems" as TraitLabel,
+				}),
+			),
+		);
+	};
+
+	public readonly getItems = (): IInventoryItem[] => {
+		return [...this._inventoryItems.values()];
+	};
+
+	protected async initializingFirstTime(): Promise<void> {
+		const legacySharedTree = this.runtime.createChannel(
+			undefined,
+			LegacySharedTree.getFactory().type,
+		) as LegacySharedTree;
+
+		const inventoryNode: BuildNode = {
+			definition: "inventory",
+			traits: {
+				// REV: If I try to remove these sample items (leaving inventoryItems as an empty array)
+				// the trait is removed entirely (undefined).  It seems that empty traits are discarded?
+				// This is a problem when later I want to iterate over it (in hasInitialized, when
+				// building up my initial set of InventoryItem objects), since I'll get a not-iterable error.
+				inventoryItems: [
+					{
+						definition: "inventoryItem",
+						traits: {
+							// REV: I tried adding an explicit "ID" trait here and using that for tracking,
```

Comment:
I think the end result of re-using node ID rather than explicitly creating a new field is probably fine here (haven't though about the details, just thinking birds-eye).

However, re: second part of the comment here: you should be able to retrieve and read information from the deleted nodes by querying the "before" TreeView (rather than "after", or your current view) passed to your viewChanged callback.

Reply:
I thought about this some more and decided a more real-world scenario would have a known "inventory ID" for each item (unrelated to node id), so I now specify an ID here.

Good pointer on using "before", I hadn't even realized it was a similar type to currentView and had completely ignored both "before" and "after" after doing the `.delta()`.  Latest revision uses this in the removal handling.